### ISSUE TITLE
这里提到的弹性训练是否一定是PS架构的，由于PS架构带宽上的限制，现在大模型的训练中使用PS架构的场景应该不多了吧？

### ISSUE THREAD

**Issue by liubowen8** [intelligent-machine-learning/dlrover#1054]

**Comment by workingloong** [comment]
TensorFlow 的 PS 和torch 的AllReduce 都支持弹性。这个文档里有弹性扩缩容的实验 https://github.com/intelligent-machine-learning/dlrover/blob/master/docs/tech_report/fault_tolerance_exps.md#scale-up-nodes。
这个里有个case，比如启动4个pod，只有3个running 了，第四个pending了，前3个可以训练。待第4个有资源的启动后，第四个也可以加入训练。相当于 torch 的弹性训练时被动扩缩容，取决于集群资源。

现在 DLRover 的主动扩缩容只支持 PS 的异步训练。因为异步训练中，主动扩缩容worker 对每次迭代的batch size 没有影响，所以我们在TF PS 异步训练做了主动扩缩容。但是torch allreduce 是同步训练，主动扩缩容worker 数量会更改每次梯度迭代 batch size = WORKER_NUM * micro-batch，对模型精度可能有影响，算法用户一般不愿意介绍。所以 torch allreduce 的弹性训练主要解决资源不够的情况下，让先启动的worker 也能训练。一般模型训练里需要根据当前的worker 数量来调整 learning rate，我们也提供了基于累积梯度方案来保证worker 扩缩容后每次迭代的batch size 不变。

**Comment by liubowen8** [comment]
你好，对于您发的这个case，我之前测试过，看到的现象是：“delete 其中一个pod，其他pod的训练进程也会退出，并重拉”。好像没有出现其他进行接续训练的弹性训练的情况。 其他三个pod的状态为running，但是内部的训练进程重拉了。
我推测是因为没有使用ps节点导致了，这个case的k8s的启动脚本中只有worker，没有ps

**Comment by liubowen8** [comment]
你好，对于您发的这个case，我之前测试过，看到的现象是：“delete 其中一个pod，其他pod的训练进程也会退出，并重拉”。好像没有出现其他进行接续训练的弹性训练的情况。 其他三个pod的状态为running，但是内部的训练进程重拉了。
我推测是因为没有使用ps节点导致了，这个case的k8s的启动脚本中只有worker，没有ps

**Comment by liubowen8** [comment]
PS 的异步训练： 基于PS的异步训练，可能导致本step使用的参数W1计算出的梯度g1，g1发送到server时，发现此时的参数已经是W2了，使用g1更新W2，会不会导致优化器更新的混乱。

**Comment by liubowen8** [comment]
“但是torch allreduce 是同步训练，主动扩缩容worker 数量会更改每次梯度迭代 batch size = WORKER_NUM * micro-batch，对模型精度可能有影响，算法用户一般不愿意接受” ：
这里对精度的影响具体是怎么形成的呢？ 是不是也可以向你们dlrover那样，相对数据进行划分，每一个DP申请使用一小块数据。如果发生故障，就把这块数据放入TO DO队列。 这些一个epoch训练下来，可以完全遍历所有的数据。这些对模型精度的影响是发生在哪一个环节呢？ 这里我没有想清楚，您看一下呢。

**Comment by liubowen8** [comment]
“但是torch allreduce 是同步训练，主动扩缩容worker 数量会更改每次梯度迭代 batch size = WORKER_NUM * micro-batch，对模型精度可能有影响，算法用户一般不愿意接受” ：
这里对精度的影响具体是怎么形成的呢？ 是不是也可以向你们dlrover那样，先对数据进行划分，每一个DP申请使用一小块数据。如果发生故障，就把这块数据放入TO DO队列。 这样一个epoch训练下来，可以完全遍历所有的数据。对模型精度的影响是发生在哪一个环节呢？ 这里我还没有想清楚，您看一下能不能解答一下呢，感谢。

**Comment by workingloong** [comment]
> 的现象是：“delete 其中一个pod，其他pod的训练进程也会退出，并重拉”。好像没有出现其他进行接续训练的弹性训练的情况。 其他三个pod的状态为running，但是内部的训练进程重拉了。
> 我推测是因为没有使用ps节点导致了，这个case的k8s的启动脚本中只有

对于 torch 是这样的。也有不重启训练进程的方案，比如 elastic horovod，但是这个方案有很多限制，我们之前基于 elastic Horovod实现过，发现一些坑。详细对比可以看这个[ https://github.com/intelligent-machine-learning
/
[dlrover](https://github.com/intelligent-machine-learning/dlrover)](https://github.com/intelligent-machine-learning/dlrover/blob/master/docs/blogs/stabilize_llm_training_cn.md#pytorch-%E5%AE%B9%E9%94%99%E6%A1%86%E6%9E%B6)

**Comment by workingloong** [comment]
> 导致本step使用的参数W1计算出的梯度g1，g1发送到server时，发现此时的参数已经是W2了，使用g1更新W2，会不会导致优化器更新的混乱。

所以 PS 异步训练主要用于搜推广的稀疏模型训练场景，这种模型有个很大的embedding，每个batch 里命中的embedding vector 重复度很低，所以梯度滞后对模型精度影响较小。

**Comment by workingloong** [comment]
> “但是torch allreduce 是同步训练，主动扩缩容worker 数量会更改每次梯度迭代 batch size = WORKER_NUM * micro-batch，对模型精度可能有影响，算法用户一般不愿意接受” ： 这里对精度的影响具体是怎么形成的呢？ 是不是也可以向你们dlrover那样，先对数据进行划分，每一个DP申请使用一小块数据。如果发生故障，就把这块数据放入TO DO队列。 这样一个epoch训练下来，可以完全遍历所有的数据。对模型精度的影响是发生在哪一个环节呢？ 这里我还没有想清楚，您看一下能不能解答一下呢，感谢。

你可以看这篇paper，https://arxiv.org/abs/1708.03888 说明了batch size 对精度的影响

**Comment by TomSuen** [comment]
您好，请问我想用dlrover启动带有deepspeed的训练脚本，目前支持吗？
"dlrover-run --network-check --exclude-straggler --nnodes=3:$NODE_NUM \
        --nproc_per_node=2 --max_restarts=3  --rdzv_conf pend_timeout=600 \
        examples/pytorch/mnist/cnn_train.py --num_epochs 5 \
        --training_data /data/mnist_png/training/ \
        --validation_data /data/mnist_png/testing/"
这里边如何改写成deepspeed的形式呢？
